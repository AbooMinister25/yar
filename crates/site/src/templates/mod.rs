mod functions;

use color_eyre::Result;
use minijinja::{Environment, context, path_loader};

use crate::{config::Config, templates::functions::pages_in_section};

const DEFAULT_404: &str = r#"
<!DOCTYPE html?
<h1> Page Not Found</h1>
<a href="{{ site.url }}">Home</a>
"#;

const DEFAULT_ATOM_FEED: &str = r#"
<?xml version="1.0" encoding="UTF-8">
<feed xmlns="http://www.w3.org/2005/Atom">
    <title> {{ site.title }} </title>
    <link href="{{ site.feed_url | safe }}" rel="self" />
    <link href="{{ site.url | safe}}"/>
    <updated> {{ last_updated | datetimeformat(format="iso") }} </updated>
    <id> {{ feed_url | safe }} </id>
    {% for page in pages %}
    <entry>
        <title> {{ page.document.frontmatter.title }} </title>
        <published> {{ page.document.date | datetimeformat(format="iso") }} </published>
        <updated> {{ page.document.updated | datetimeformat(format="iso") }} </updated>
        <id> {{ page.permalink | safe }} </id>
        {% if page.site.authors %}
            {% for author in page.authors %}
            <author>
                <name> {{ page.site.author }} </name>
            </author
            {% endfor %}
        {% else %}
            <author>
                <name> Unknown </name>
            </author>
        {% endif %}
        <summary> {{ page.document.summary }} </summary>
        <content type="html">
            {{ page.document.content }}
        </content>
    </entry>
    {% endfor %}
</feed>
"#;

const DEFAULT_SITEMAP: &str = r#"
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
    {%- for page in pages %}
    <url>
        <loc>{{ page.permalink | escape | safe }}</loc>
        <lastmod>{{ page.document.updated }}</lastmod>
    </url>
    {%- endfor %}
</urlset>
"#;

pub fn create_environment(config: &Config) -> Result<Environment<'static>> {
    let mut env = Environment::new();
    env.add_template("404.html", DEFAULT_404)?;
    env.add_template("atom.xml", DEFAULT_ATOM_FEED)?;
    env.add_template("sitemap.xml", DEFAULT_SITEMAP)?;
    env.set_loader(path_loader(&config.root.join("templates")));
    env.add_global(
        "site",
        context! { site => context!{
            url => config.url,
            authors => config.authors,
            title => config.title,
            description => config.description,
        }},
    );
    env.add_function("pages_in_section", pages_in_section);
    minijinja_contrib::add_to_environment(&mut env);

    Ok(env)
}
