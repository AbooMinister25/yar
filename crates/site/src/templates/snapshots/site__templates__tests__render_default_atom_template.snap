---
source: crates/site/src/templates/mod.rs
expression: rendered
---
"\n<?xml version=\"1.0\" encoding=\"UTF-8\">\n<feed xmlns=\"http://www.w3.org/2005/Atom\">\n    <title> none </title>\n    <link href=\"http://0.0.0.0:8000/atom.xml\" rel=\"self\" />\n    <link href=\"http://0.0.0.0:8000/\"/>\n    <updated> 2025-01-01T00:01:01+00:00 </updated>\n    <id> http://0.0.0.0:8000/atom.xml </id>\n    \n    \n    <entry>\n        <title> post-0 </title>\n        <published> 2025-01-01T06:00:00+00:00 </published>\n        <updated> 2025-03-12T08:00:00+00:00 </updated>\n        <id> https://example.com/series/testing/post-0 </id>\n        \n            <author>\n                <name> Unknown </name>\n            </author>\n        \n        <summary> <p>Hello World</p>\n </summary>\n        <content type=\"html\">\n            <p>Hello World</p>\n\n        </content>\n    </entry>\n    \n    \n    \n    <entry>\n        <title> post-1 </title>\n        <published> 2025-01-01T06:00:00+00:00 </published>\n        <updated> 2025-03-12T08:00:00+00:00 </updated>\n        <id> https://example.com/series/testing/post-1 </id>\n        \n            <author>\n                <name> Unknown </name>\n            </author>\n        \n        <summary> <p>Hello World</p>\n </summary>\n        <content type=\"html\">\n            <p>Hello World</p>\n\n        </content>\n    </entry>\n    \n    \n    \n    <entry>\n        <title> post-2 </title>\n        <published> 2025-01-01T06:00:00+00:00 </published>\n        <updated> 2025-03-12T08:00:00+00:00 </updated>\n        <id> https://example.com/series/testing/post-2 </id>\n        \n            <author>\n                <name> Unknown </name>\n            </author>\n        \n        <summary> <p>Hello World</p>\n </summary>\n        <content type=\"html\">\n            <p>Hello World</p>\n\n        </content>\n    </entry>\n    \n    \n    \n    <entry>\n        <title> post-3 </title>\n        <published> 2025-01-01T06:00:00+00:00 </published>\n        <updated> 2025-03-12T08:00:00+00:00 </updated>\n        <id> https://example.com/series/testing/post-3 </id>\n        \n            <author>\n                <name> Unknown </name>\n            </author>\n        \n        <summary> <p>Hello World</p>\n </summary>\n        <content type=\"html\">\n            <p>Hello World</p>\n\n        </content>\n    </entry>\n    \n    \n    \n    <entry>\n        <title> post-4 </title>\n        <published> 2025-01-01T06:00:00+00:00 </published>\n        <updated> 2025-03-12T08:00:00+00:00 </updated>\n        <id> https://example.com/series/testing/post-4 </id>\n        \n            <author>\n                <name> Unknown </name>\n            </author>\n        \n        <summary> <p>Hello World</p>\n </summary>\n        <content type=\"html\">\n            <p>Hello World</p>\n\n        </content>\n    </entry>\n    \n    \n    \n    <entry>\n        <title> post-5 </title>\n        <published> 2025-01-01T06:00:00+00:00 </published>\n        <updated> 2025-03-12T08:00:00+00:00 </updated>\n        <id> https://example.com/series/testing/post-5 </id>\n        \n            <author>\n                <name> Unknown </name>\n            </author>\n        \n        <summary> <p>Hello World</p>\n </summary>\n        <content type=\"html\">\n            <p>Hello World</p>\n\n        </content>\n    </entry>\n    \n    \n    \n    <entry>\n        <title> post-6 </title>\n        <published> 2025-01-01T06:00:00+00:00 </published>\n        <updated> 2025-03-12T08:00:00+00:00 </updated>\n        <id> https://example.com/series/testing/post-6 </id>\n        \n            <author>\n                <name> Unknown </name>\n            </author>\n        \n        <summary> <p>Hello World</p>\n </summary>\n        <content type=\"html\">\n            <p>Hello World</p>\n\n        </content>\n    </entry>\n    \n    \n    \n    <entry>\n        <title> post-7 </title>\n        <published> 2025-01-01T06:00:00+00:00 </published>\n        <updated> 2025-03-12T08:00:00+00:00 </updated>\n        <id> https://example.com/series/testing/post-7 </id>\n        \n            <author>\n                <name> Unknown </name>\n            </author>\n        \n        <summary> <p>Hello World</p>\n </summary>\n        <content type=\"html\">\n            <p>Hello World</p>\n\n        </content>\n    </entry>\n    \n    \n    \n    <entry>\n        <title> post-8 </title>\n        <published> 2025-01-01T06:00:00+00:00 </published>\n        <updated> 2025-03-12T08:00:00+00:00 </updated>\n        <id> https://example.com/series/testing/post-8 </id>\n        \n            <author>\n                <name> Unknown </name>\n            </author>\n        \n        <summary> <p>Hello World</p>\n </summary>\n        <content type=\"html\">\n            <p>Hello World</p>\n\n        </content>\n    </entry>\n    \n    \n    \n    <entry>\n        <title> post-9 </title>\n        <published> 2025-01-01T06:00:00+00:00 </published>\n        <updated> 2025-03-12T08:00:00+00:00 </updated>\n        <id> https://example.com/series/testing/post-9 </id>\n        \n            <author>\n                <name> Unknown </name>\n            </author>\n        \n        <summary> <p>Hello World</p>\n </summary>\n        <content type=\"html\">\n            <p>Hello World</p>\n\n        </content>\n    </entry>\n    \n    \n</feed>"
